### ISSUE TITLE
Windows: Implement a way to enable color in console even when the output is redirected

### ISSUE THREAD

**Issue by NickCraver** [dotnet/runtime#68340]
This is a follow-up to #33980 which was closed as fixed by #47935, but only for Unix/Linux, because Unix and Windows use a different PAL for the tail end of console. The ultimate feature was enabling a new environmental variable: `DOTNET_SYSTEM_CONSOLE_ALLOW_ANSI_COLOR_REDIRECTION`, which when set to `1` or `true` will still emit color codes when the console is redirected (e.g. under CI conditions for logging).

Here's the Unix PAL implementation (also respecting `NO_COLOR`): https://github.com/dotnet/runtime/blob/ed04fef3b744ad179f4c570b873721d09534c017/src/libraries/System.Console/src/System/ConsolePal.Unix.cs#L874

And here's the Windows PAL: https://github.com/dotnet/runtime/blob/ed04fef3b744ad179f4c570b873721d09534c017/src/libraries/System.Console/src/System/ConsolePal.Windows.cs

Having color output in environments like GitHub Actions is immensely helpful when finding errors in massive logs for example, compare these Unix and Windows runs today ([here's an example PR](https://github.com/StackExchange/StackExchange.Redis/pull/2113), though the logs will likely expire long-term):

<img width="686" alt="Windows Build (no color)" src="https://user-images.githubusercontent.com/454813/164516979-c0612481-e820-475b-806b-d9b3704a8244.png">

<img width="806" alt="Linux Build (color!)" src="https://user-images.githubusercontent.com/454813/164517144-13d8d7e4-f17d-498d-bbe8-27b580ea742e.png">

The ask is: **can we please make `DOTNET_SYSTEM_CONSOLE_ALLOW_ANSI_COLOR_REDIRECTION` work on Windows as well?**

cc @0xced @Tyrrrz @stephentoub

**Comment by stephentoub** [comment]
> The ask is: can we please make DOTNET_SYSTEM_CONSOLE_ALLOW_ANSI_COLOR_REDIRECTION work on Windows as well?

This isn't really about making DOTNET_SYSTEM_CONSOLE_ALLOW_ANSI_COLOR_REDIRECTION "work".  This is about changing Console on Windows to use ANSI escape sequences rather than using Win32 API calls as it does today.

**Comment by NickCraver** [comment]
@stephentoub yeah I agree based on what I'm aware of here (admitted novice though). From the user standpoint though, that'd be the functionality/surface still to maintain existing behavior right? We wouldn't want to suddenly light up color output in all scenarios and need some control - ultimately I'm proposing the same control when this is feasible. All I know is with online tools (Actions, DevOps, etc.) I'm searching through a _lot_ of logs spending and a _lot_ of time in oh so many builds...it'd be a huge human time saver for anyone in similar positions if we could colorize.

**Comment by stephentoub** [comment]
I'm not arguing for or against the scenario or its worth, just highlighting that this isn't simply a matter of flipping a switch or adding an if block to respond to the environment variable: it requires fundamentally changing what `Console.ForegroundColor/BackgroundColor/ResetColor` do.

**Comment by xt0rted** [comment]
This explains why I was able to get my dotnet local tools outputting color on GitHub Actions but not the dotnet cli. I'm using the helpers in [`System.CommandLine.Rendering`](https://github.com/dotnet/command-line-api/blob/dd44dd05a20bf6af02615d8363948a630e821d00/src/System.CommandLine.Rendering/Ansi.cs) which writes to `Console.Out`. I think the [ansi support check](https://github.com/dotnet/command-line-api/blob/dd44dd05a20bf6af02615d8363948a630e821d00/src/System.CommandLine.Rendering/ConsoleFormatInfo.cs#L96-L111) will need updating too to force color output on Windows.

**Comment by xt0rted** [comment]
I was seeing differing results locally vs. on github so I dug into it a bit more. The way the dotnet cli emits color codes doesn't work with how the actions build log parses them. Making this change will get us color on windows, but the cli output won't give us the full experience unless github changes the build log parser, or the cli changes how colors are emitted. I opened [a discussion](https://github.com/github/feedback/discussions/15416) about this but the short of it is the build log doesn't show colors when they span multiple lines.

**Comment by heaths** [comment]
Maybe tangential - and quite likely a separate feature request - but what about adding `Console.IsColorEnabled` or something and allowing developers to decide whether they want to emit ANSI color codes or not. For built-in console hosts on Windows, this is only a problem in conhost and even then only if virtual terminal sequences weren't enabled. It's been pretty common to enable in third party libraries e.g., golang modules or rustlang crates, since they're often used for CLIs. Providing a capability like that can lead to all sorts of console color theming capabilities / libraries.

As to "how much color" e.g., 16, 256, or truecolor, I've been in a few discussions with @dhowett from the console / Windows Terminal team about this e.g., https://github.com/microsoft/terminal/issues/11057. Since Unix's `ConsolePal` already loads terminfo it's straight forward to know how many colors are likely supported, if any. On Windows, there's some assumptions than can be made per the aforementioned thread. IIRC, if you can enable VT sequences, truecolor is supported. To enable VT sequences, e.g. in C/C++: https://github.com/microsoft/vswhere/blob/c7a417bfb46cd6a1c3e31a91ba51193d7659b6f8/src/vswhere.lib/Console.cpp#L116-L128

**Comment by baronfel** [comment]
MSBuild Terminal Logger and dotnet/test ANSI output detection would both make use of IsColorEnabled and in general exposing the TermInfo knowledge that the runtime has.

**Comment by am11** [comment]
> As to "how much color" e.g., 16, 256, or truecolor

It feels like some/most of the use-cases requiring `IsColorEnabled` are due to the 16 colors limitation of `ConsoleColor`, for which there is a default 'without color' redirection behavior provided OOTB that can be altered by `DOTNET_SYSTEM_CONSOLE_ALLOW_ANSI_COLOR_REDIRECTION` and `NO_COLOR` environment variables. For more colors, we need to resort to escape sequences (part of the string content), which get ignored by those env vars mechanism.

Perhaps, the proposal in https://github.com/dotnet/runtime/issues/39746 could be extended to support RGB for truecolor:

```diff
+ public static void WriteLine(ConsoleColor foreground, ConsoleColor background, string value) { throw null; }
+ public static void WriteLine((byte R, byte G, byte B) foreground, (byte R, byte G, byte B) background, bool is24BitColor, string value) { throw null; }

+ public static void Write(ConsoleColor foreground, ConsoleColor background, string value) { throw null; }
+ public static void Write((byte R, byte G, byte B) foreground, (byte R, byte G, byte B) background, bool is24BitColor, string value) { throw null; }
```

map to:

```c#
{
    if (shouldUseColors)
    {
        if (is24BitColor) // 256color
        {
            value = $"\e[38;5;{GetColorCode(foreground.R, foreground.G, foreground.B)}m\e[48;5;{GetColorCode(background.R, background.G, background.B)}m{value}\e[0m";

            static int GetColorCode(byte R, byte G, byte B) => 16 + (36 * (R / 51)) + (6 * (G / 51)) + (B / 51);
        }
        else // truecolor
        {
            value = $"\e[38;2;{foreground.R};{foreground.G};{foreground.B}m\e[48;2;{background.R};{background.G};{background.B}m{value}\e[0";
        }
    }

    Write(value);
}
```

This way, use-cases requiring `IsColorEnabled`-type details for handling beyond the 16-color limit will be handled more idiomatically.
